packages/target-decisioning-engine/src/geoProvider.js

```diff
@@ -0,0 +1,79 @@
+import {
+  getFetchApi,
+  isDefined,
+  isUndefined,
+  isValidIpAddress
+} from "@adobe/target-tools";
+
+import { getGeoLookupPath } from "./utils";
+
+/**
+ * @param { Response } fetchResponse;
+ * @return { import("@adobe/target-tools/delivery-api-client/models/Geo").Geo }
+ */
+export function createGeoObject(fetchResponse) {
+  const geoObject = {};
+
+  const ipAddress = fetchResponse.headers.get("X-Forwarded-For");
+  if (ipAddress != null && isDefined(ipAddress)) {
+    geoObject.ipAddress = ipAddress;
+  }
+  const latitude = fetchResponse.headers.get("X-GEO-Latitude");
+  if (latitude != null && isDefined(latitude)) {
+    geoObject.latitude = parseFloat(latitude);
+  }
+  const longitude = fetchResponse.headers.get("X-GEO-Longitude");
+  if (longitude != null && isDefined(longitude)) {
+    geoObject.longitude = parseFloat(longitude);
+  }
+  const countryCode = fetchResponse.headers.get("X-GEO-Country-Code");
+  if (countryCode != null && isDefined(countryCode)) {
+    geoObject.countryCode = countryCode;
+  }
+  const stateCode = fetchResponse.headers.get("X-GEO-Region-Code");
+  if (stateCode != null && isDefined(stateCode)) {
+    geoObject.stateCode = stateCode;
+  }
+  const city = fetchResponse.headers.get("X-GEO-City");
+  if (city != null && isDefined(city)) {
+    geoObject.city = city;
+  }
+  return geoObject;
+}
+
+/**
+ * @param {import("../types/DecisioningConfig").DecisioningConfig} config
+ * @param { import("../types/DecisioningArtifact").DecisioningArtifact } artifact
+ */
+export function GeoProvider(config, artifact) {
+  const fetchApi = getFetchApi(config.fetchApi);
+  const { geoTargetingEnabled = false } = artifact;
+
+  /**
+   * @param {import("@adobe/target-tools/delivery-api-client/models/Geo").Geo} geoRequestContext
+   * @return { import("@adobe/target-tools/delivery-api-client/models/Geo").Geo }
+   */
+  async function validGeoRequestContext(geoRequestContext = {}) {
+    // When ipAddress is the only geo value passed in to getOffers(), do IP-to-Geo lookup.
+    const geoLookupPath = getGeoLookupPath(config);
+
+    if (
+      geoTargetingEnabled &&
+      isValidIpAddress(geoRequestContext.ipAddress) &&
+      isUndefined(geoRequestContext.latitude) &&
+      isUndefined(geoRequestContext.longitude) &&
+      isUndefined(geoRequestContext.countryCode) &&
+      isUndefined(geoRequestContext.stateCode) &&
+      isUndefined(geoRequestContext.city)
+    ) {
+      const ipToGeoLookupResponse = await fetchApi(geoLookupPath);
```

Comment:
@mptap This request is not correct.  It needs to include the `x-forwarded-for` header with the `geoRequestContext.ipAddress`.  The way it is now, it will get the geo location for the ip address of the web server, not the ip address specified in the request.

Reply:
@jasonwaters I am already parsing the `x-forwarded-for` header for `geoRequestContext.ipAddress`. See line 17 of  this file.

Reply:
Updated